### PR TITLE
Add Support for VertexAI Gemini

### PR DESCRIPTION
**Pull request by ajac-zero**
## Vertex AI x Instructor

This is an initial integration of the Vertex AI SDK so that instructor works with Gemini models via **tool calling**.

Addresses #592, #441, and others (like myself) who want to use Gemini (Vertex AI version) with instructor!

### Usage
```python
import vertexai.generative_models as gm
from pydantic import BaseModel, Field
import instructor
import vertexai

class Item(BaseModel):
    name: str
    price: float = Field(description="If missing, set to 0")

class Order(BaseModel):
    items: list[Item]
    customer: str

vertexai.init()

model = gm.GenerativeModel("gemini-1.5-pro-preview-0409")

gemini = instructor.from_vertexai(model)

model = gemini.create(
    response_model=Order,
    messages=[{"role": "user", "content": """
    Library visit details:
    Visitor: Jason
    Library ID: LIB123456
    Books checked out:
    - Title: The Great Adventure, Author: Jane Doe, Genre: Fantasy, ISBN: 1234567890
    - Title: History of Tomorrow, Author: John Smith, Genre: Non-Fiction, ISBN: 0987654321
    """}]
)

print(model)
# >>> output
# Order(
#     items=[
#         Item(name='The Great Adventure', price=0.0),
#         Item(name='History of Tomorrow', price=0.0)
#         ],
#     customer='Jason'
#     )
```

### Note: Gemini tool calling is still in preview

Gemini does not support the entire OpenAPI Schema yet, this is straight from their docs:

> Vertex AI offers limited support of the OpenAPI schema. The following attributes are supported: type, nullable, required, format, description, properties, items, enum. The following attributes are not supported: default, optional, maximum, oneOf.

So some pydantic fields like default_factory, or even some standard types like Unions will raise exceptions. Though hopefully this will resolve itself without any changes as Gemini start supporting more attributes. (That's why in the example I tell Gemini to place a 0 if the price is missing, as trying `Union[float, None]` raises an exception.)

----

| :rocket: This description was created by [Ellipsis](https://www.ellipsis.dev) for commit c5c488eb65277e20a8d5607f02de2050d719e1ab  | 
|--------|

### Summary:
This PR integrates VertexAI Gemini models into the Instructor library, adding necessary client code, tests, and dependencies.

**Key points**:
- Added support for VertexAI Gemini models in the Instructor library.
- Created a new client file `client_vertexai.py` for VertexAI integration.
- Updated `instructor/__init__.py` to conditionally import VertexAI client based on package availability.
- Added new test cases in `tests/llm/test_vertexai/` to cover VertexAI functionality.
- Updated project dependencies in `pyproject.toml` to include VertexAI-related packages.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -87,6 +89,9 @@ litellm = "^1.35.31"
 [tool.poetry.group.google-generativeai.dependencies]
 google-generativeai = "^0.5.4"
 
+[tool.poetry.group.vertexai.dependencies]
```
[the remark this answers is not in the document]
Comment (ajac-zero) [reply pyproject.toml:93]: fixed 👍

**Comment by jxnl** [comment]
can you make sure to update the patch.md docs to cover the gotchas

**Comment by asm** [comment]
I hate to delay this PR any further since I'm really looking forward to using Instructor.  However, it looks like [1.53.0](https://github.com/googleapis/python-aiplatform/releases/tag/v1.53.0) was *just* released with support for setting the response mime type and response schema 😄

**Comment by asm** [comment]
I should add, I don't think you need to use a function/tool to get valid JSON back.  You could just use something like:
```
config = GenerationConfig(
  response_mime_type = 'application/json',
  response_schema = model.schema(),
)
response = client.generate_content(
  contents = ...,
  generation_config = config,
)
```

**Comment by jxnl** [comment]
does this conflict with #726  #719 


I see these a few PRs

**Comment by ajac-zero** [comment]
I'm fairly confident there is no overlap. All the logic for vertex ai is behind VERTEXAI_TOOLS mode, which doesn't share logic with the gemini modes.

**Comment by punjabdhaputar** [comment]
Will this work for VertexAI with litellm? I am currently running VertexAI with litellm and running into this issue: 
```
APIError: VertexAIException - Unknown field for Schema: $defs 
Model: gemini-1.5-pro-preview-0514
Messages: [{'role': 'system', 'content'
```

**Comment by ajac-zero** [comment]
@punjabdhaputar It will not, this is a compatibility issue between pydantic  and the current gemini model.

When we use `model_json_schema()` on a pydantic model with nested models, we get an OpenAPI schema like this:

```python
from pydantic import BaseModel

class Address(BaseModel):
    street: str
    city: str

class User(BaseModel):
    name: str
    address: Address

schema = User.model_json_schema()

print(schema)
# {
#     "$defs": {
#         "Address": {
#             "properties": {
#                 "street": {
#                     "title": "Street",
#                     "type": "string"
#                 },
#                 "city": {
#                     "title": "City",
#                     "type": "string"
#                 }
#             },
#             "required": ["street", "city"],
#             "title": "Address",
#             "type": "object"
#         }
#     },
#     "properties": {
#         "name": {
#             "title": "Name",
#             "type": "string"
#         },
#         "address": {
#             "$ref": "#/$defs/Address"
#         }
#     },
#     "required": ["name", "address"],
#     "title": "User",
#     "type": "object"
# }
```
Pydantic first uses `$defs` to define the nested model, and then inserts the definition into the main model with `"$ref": "#/$defs/Address"`. Gemini **DOES NOT** support this type of nested definition (AFAIK OpenAI and Anthropic do), so it raises an exception.

In the current Gemini implementation, we get around this by replacing the reference with the definition directly using `jsonref`. Litellm would have to do something similar to ensure interoperability with the other providers.

This is also relevant in issue #748

**Comment by punjabdhaputar** [comment]
Thank you for the detailed response! I can follow-up with litellm in case there is anything they can do on that side!

**Comment by osiloke** [comment]
Hi, it seems you already support gemini json, would it be possible to just allow this.

```
litellm_client = instructor.from_litellm(
    completion,
    mode=instructor.Mode.GEMINI_JSON,
)
```
So that the response is handled as a gemini tool json spec.

Currently you can only use the GEMINI_JSON mode with just gemini.

**Comment by gokturkDev** [comment]
I am getting   
```
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/ai/models/gemini_model.py", line 37, in get_response_with_response_model
    |     return self.instructor_model.create(
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/instructor/client.py", line 91, in create
    |     return self.create_fn(
    |            ^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/instructor/patch.py", line 143, in new_create_sync
    |     response = retry_sync(
    |                ^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/instructor/retry.py", line 151, in retry_sync
    |     for attempt in max_retries:
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/tenacity/__init__.py", line 435, in __iter__
    |     do = self.iter(retry_state=retry_state)
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/tenacity/__init__.py", line 368, in iter
    |     result = action(retry_state)
    |              ^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/tenacity/__init__.py", line 410, in exc_check
    |     raise retry_exc.reraise()
    |           ^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/tenacity/__init__.py", line 183, in reraise
    |     raise self.last_attempt.result()
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    |     return self.__get_result()
    |            ^^^^^^^^^^^^^^^^^^^
    |   File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    |     raise self._exception
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/instructor/retry.py", line 154, in retry_sync
    |     response = func(*args, **kwargs)
    |                ^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/vertexai/generative_models/_generative_models.py", line 524, in generate_content
    |     return self._generate_content(
    |            ^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/vertexai/generative_models/_generative_models.py", line 613, in _generate_content
    |     gapic_response = self._prediction_client.generate_content(request=request)
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/google/cloud/aiplatform_v1beta1/services/prediction_service/client.py", line 2125, in generate_content
    |     response = rpc(
    |                ^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    |     return wrapped_func(*args, **kwargs)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/venv/lib/python3.12/site-packages/google/api_core/grpc_helpers.py", line 78, in error_remapped_callable
    |     raise exceptions.from_grpc_error(exc) from exc
    | google.api_core.exceptions.ResourceExhausted: 429 Quota exceeded for aiplatform.googleapis.com/generate_content_requests_per_minute_per_project_per_base_model with base model: gemini-1.5-flash. Please submit a quota increase request. https://cloud.google.com/vertex-ai/docs/generative-ai/quotas-genai.
    +------------------------------------

```
errors when using 

```
model = gemini.create(
    response_model=Order,
    messages=[{"role": "user", "content": """
    Library visit details:
    Visitor: Jason
    Library ID: LIB123456
    Books checked out:
    - Title: The Great Adventure, Author: Jane Doe, Genre: Fantasy, ISBN: 1234567890
    - Title: History of Tomorrow, Author: John Smith, Genre: Non-Fiction, ISBN: 0987654321
    """}]
)

```

When using it without instructor like this 
```
vertexai.init(project=project_id, location="us-central1")

model = GenerativeModel(model_name="gemini-1.5-flash-001")

response = model.generate_content(
    "What's a good name for a flower shop that specializes in selling bouquets of dried flowers?"
)
```

Everything works fine, which I believe means that the problem is not in the quota per se. Could you provide me a direction to go with this error?

**Comment by osiloke** [comment]
> I am getting
> 
> ```
>     |   File "/Users/gokturkdev/PycharmProjects/hekades-backend/ai/models/gemini_model.py", line 37, in get_response_with_response_model
>     |     ...
>     +------------------------------------
> ```
> 
> errors when using
> 
> ```
> model = gemini.create(
>   ...
> Everything works fine, which I believe means that the problem is not in the quota per se. Could you provide me a direction to go with this error?

Hey, you seem to be hitting one of gemini's api quota limits. You can also try my [integration steps](https://github.com/jxnl/instructor/pull/711#issuecomment-2185036728).

**Comment by ajac-zero** [comment]
@gokturkDev I couldn't replicate the issue, but what is most likely happening is that the quota limit is being hit during the retries to fix the model output.

Gemini can often get the model wrong first try, usually due to types (ej. outputting ints when the model needs floats). If the generated model is wrong we immediately respond with the validation error and Gemini tries again with the additional info, by default this is repeated up to 3 times.

So you are most likely hitting the limit with these retry messages. You could try using `VERTEXAI_JSON` mode instead and see if you are able to get better results, or try to simplify the model if possible to lower the amount of retries needed, if at all.

**Pull request merged**